src/IssueViz.Security/OpenInIDE/Api/IOpenInIDEStateValidator.cs

```diff
@@ -41,4 +47,114 @@ internal interface IOpenInIDEStateValidator
+            ILogger logger)
+        {
+            this.infoBarManager = infoBarManager;
+            this.configurationProvider = configurationProvider;
+            this.outputWindowService = outputWindowService;
+            this.logger = logger;
+        }
+
+        public bool CanHandleOpenInIDERequest(Uri serverUri, string projectKey, string organizationKey)
+        {
+            RemoveExistingInfoBar();
+
+            var failureMessage = GetFailureMessage(serverUri, projectKey, organizationKey);
+
+            if (string.IsNullOrEmpty(failureMessage))
+            {
+                return true;
+            }
+
+            AddInfoBar();
+            logger.WriteLine(failureMessage);
+
+            return false;
+        }
+
+        private string GetFailureMessage(Uri serverUri, string projectKey, string organizationKey)
+        {
+            var configuration = configurationProvider.GetConfiguration();
+            string reason = null;
+
+            if (configuration.Mode == SonarLintMode.Standalone)
+            {
+                reason = OpenInIDEResources.RequestValidator_InvalidStateReason_NotInConnectedMode;
+            }
+            else if (!configuration.Project.ServerUri.Equals(serverUri))
+            {
+                reason = string.Format(OpenInIDEResources.RequestValidator_InvalidStateReason_WrongServer, configuration.Project.ServerUri);
+            }
+            else if (!string.IsNullOrEmpty(organizationKey) &&
+                (configuration.Project.Organization == null ||
+                 !organizationKey.Equals(configuration.Project.Organization.Key, StringComparison.OrdinalIgnoreCase)))
+            {
+                reason = string.Format(OpenInIDEResources.RequestValidator_InvalidStateReason_WrongOrganization, configuration.Project.Organization?.Key);
+            }
+            else if (!configuration.Project.ProjectKey.Equals(projectKey, StringComparison.OrdinalIgnoreCase))
+            {
+                reason = string.Format(OpenInIDEResources.RequestValidator_InvalidStateReason_WrongProject, configuration.Project.ProjectKey);
+            }
+
+            if (reason == null)
+            {
+                return null;
+            }
+
+            var instructions = string.IsNullOrEmpty(organizationKey)
+                ? string.Format(OpenInIDEResources.RequestValidator_Instructions_SonarQube, serverUri, projectKey)
+                : string.Format(OpenInIDEResources.RequestValidator_Instructions_SonarCloud, serverUri, organizationKey, projectKey);
+
+            var fullMessage = string.Format(OpenInIDEResources.RequestValidator_InvalidState, reason, instructions);
+
+            return fullMessage;
+        }
+
+        private void AddInfoBar()
+        {
+            currentInfoBar = infoBarManager.AttachInfoBarWithButton(new Guid(HotspotsToolWindow.ToolWindowId), OpenInIDEResources.RequestValidator_InfoBarMessage, "Show Output Window", default);
```

Comment:
Could assert that the currentInfoBar is null to start with.

Reply:
Ah, my bad, apologies. I missed the PR comment.